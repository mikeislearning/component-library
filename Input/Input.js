import React from 'react';
import { Image, TextInput, TouchableOpacity, View } from 'react-native';
import { PropTypes } from 'prop-types';

import { images } from 'app/theme';
import { FormError } from 'app/components';

import colors from '../../theme/colors';

import styles from './Input.styles';

export default class Input extends React.Component {
  static propTypes = {
    clearMode: PropTypes.string,
    customStyle: PropTypes.object,
    errorMsg: PropTypes.string,
    errors: PropTypes.bool,
    hasBorder: PropTypes.bool,
    hasBottomMargin: PropTypes.bool,
    onChangeText: PropTypes.func,
    onPressReveal: PropTypes.func,
    multiline: PropTypes.bool,
    placeholder: PropTypes.string,
    textContentType: PropTypes.string,
    value: PropTypes.string,
  };

  static defaultProps = {
    clearMode: 'always',
    customStyle: {},
    hasBorder: true,
    hasBottomMargin: true,
  };

  render() {
    const {
      clearMode,
      customStyle,
      errorMsg,
      errors,
      hasBorder,
      hasBottomMargin,
      multiline,
      onChangeText,
      onPressReveal,
      placeholder,
      textContentType,
      value,
      ...otherProps
    } = this.props;

    const inputStyle = [styles.input, customStyle];
    const containerStyle = [];

    if (hasBorder && !multiline) inputStyle.push(styles.hasBorder);
    if (multiline) inputStyle.push(styles.multiline);
    if (!value) inputStyle.push(styles.inputEmpty);
    if (errors) inputStyle.push(styles.errorInput);

    if (hasBottomMargin) containerStyle.push(styles.hasBottomMargin);
    if (multiline) {
      containerStyle.push(styles.multilineContainer);
      if (hasBorder) containerStyle.push(styles.hasBorder);
      if (!value) containerStyle.push(styles.inputEmpty);
    }

    return (
      <View style={containerStyle}>
        <View style={styles.inputContainer}>
          <TextInput
            clearButtonMode={clearMode}
            onChangeText={onChangeText}
            value={value}
            placeholder={placeholder}
            style={inputStyle}
            multiline={multiline}
            placeholderTextColor={colors.gray.medium}
            {...otherProps}
          />
          {textContentType === 'password' && (
            <TouchableOpacity
              accessibilityLabel="Reveal password"
              activeOpacity={0.8}
              onPress={onPressReveal}
              style={styles.iconContainer}
            >
              <Image source={images.eye} style={styles.icon} />
            </TouchableOpacity>
          )}
        </View>
        {errorMsg && <FormError text={errorMsg} />}
      </View>
    );
  }
}
