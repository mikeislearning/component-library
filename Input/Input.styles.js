import { StyleSheet } from 'react-native';

import { colors, fonts } from 'app/theme';

const styles = StyleSheet.create({
  errorInput: {
    borderColor: colors.red.light,
  },
  hasBorder: {
    backgroundColor: colors.white,
    borderColor: colors.black,
    borderWidth: 1,
  },
  hasBottomMargin: {
    marginBottom: 10,
  },
  icon: {
    height: 12,
    width: 18,
  },
  iconContainer: {
    paddingHorizontal: 17,
    paddingVertical: 19,
    position: 'absolute',
    right: 0,
    top: 0,
  },
  input: {
    backgroundColor: colors.white,
    borderRadius: 8,
    color: colors.black,
    fontFamily: fonts.family.messinaSans.regular,
    fontSize: 16,
    height: 50,
    paddingBottom: 0,
    paddingHorizontal: 10,
    paddingTop: 0,
    textAlignVertical: 'center',
  },
  inputContainer: {
    position: 'relative',
  },
  inputEmpty: {
    backgroundColor: colors.gray.light,
    borderColor: colors.gray.medium,
    fontFamily: fonts.family.messinaSans.lightItalic,
  },
  multiline: {
    height: 100,
    lineHeight: 22,
    paddingHorizontal: 14,
    paddingVertical: 15,
    textAlignVertical: 'top',
  },
  multilineContainer: {
    backgroundColor: colors.white,
    borderRadius: 8,
    paddingBottom: 10,
    paddingHorizontal: 1,
    paddingTop: 15,
  },
});

export default styles;
