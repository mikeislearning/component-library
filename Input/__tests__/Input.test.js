/**
 * @format
 */

// import 'react-native';
import React from 'react';
import { fireEvent, render } from 'react-native-testing-library';

import Input from '../Input';

describe('<Input />', () => {
  it('renders correctly', () => {
    const input = render(<Input />);
    expect(input).toMatchSnapshot();
  });

  it('can enter input successfully', () => {
    const onEventMock = jest.fn();
    const { getByTestId } = render(<Input testID="change" onChangeText={onEventMock} />);

    // fireEvent(getByTestId('change'), 'onChangeText', 'testing123');
    fireEvent.changeText(getByTestId('change'), 'testing123');
    expect(onEventMock).toHaveBeenCalledWith('testing123');
  });
});
