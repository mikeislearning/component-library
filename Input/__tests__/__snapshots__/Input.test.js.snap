// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      clearButtonMode="always"
      placeholderTextColor="#726E67"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 8,
            "color": "#000000",
            "fontFamily": "MessinaSans-Regular",
            "fontSize": 16,
            "height": 50,
            "minWidth": 260,
            "paddingBottom": 0,
            "paddingHorizontal": 10,
            "paddingTop": 0,
            "textAlignVertical": "center",
          },
          Object {},
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#000000",
            "borderWidth": 1,
          },
          Object {
            "backgroundColor": "#ECECEC",
            "borderColor": "#726E67",
            "fontFamily": "MessinaSans-LightItalic",
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
